lib/src/podman_ostree.rs

```diff
@@ -0,0 +1,272 @@
+                move_to_trash(&pathbuf)?;
+            }
+        }
+        // Create the directory if needed.
+        if needs_create {
+            let mut db = DirBuilder::new();
+            db.mode(src_meta.mode());
+            output
+                .create_dir_with(&pathbuf, &db)
+                .with_context(|| format!("Creating {pathbuf:?}"))?;
+        }
+        // Now recurse
+        merge_layer(layer, pathbuf, output, state)?;
+    } else if (src_meta.mode() & libc::S_IFMT) == libc::S_IFCHR && src_meta.rdev() == 0 {
+        // The layer specifies a whiteout entry; remove the target path.
+        if target_meta.is_some() {
+            move_to_trash(&pathbuf)?;
+        }
+    } else {
+        // We're operating on a non-directory.  In this case if the target exists,
+        // it needs to be removed.
+        if target_meta.is_some() {
+            move_to_trash(&pathbuf)?;
+        }
+        if src_meta.is_symlink() {
+            let target =
+                cap_primitives::fs::read_link_contents(&layer.as_filelike_view(), &pathbuf)
+                    .with_context(|| format!("Reading link {pathbuf:?}"))?;
+            cap_primitives::fs::symlink_contents(target, &output.as_filelike_view(), &pathbuf)
+                .with_context(|| format!("Writing symlink {pathbuf:?}"))?;
+        } else {
+            let src = layer
+                .open(&pathbuf)
+                .with_context(|| format!("Opening src {pathbuf:?}"))?;
+            // Use reflinks if available, otherwise we can fall back to hard linking.  The hardlink
+            // count will "leak" into any containers spawned (until podman learns to use composefs).
+            if state.can_clone {
+                let mut openopts = cap_std::fs::OpenOptions::new();
+                openopts.write(true);
+                openopts.create_new(true);
+                openopts.mode(src_meta.mode());
+                let dest = output
+                    .open_with(&pathbuf, &openopts)
+                    .with_context(|| format!("Opening dest {pathbuf:?}"))?;
+                rustix::fs::ioctl_ficlone(dest.as_fd(), src.as_fd()).context("Cloning")?;
+            } else {
+                layer
+                    .hard_link(&pathbuf, output, &pathbuf)
+                    .context("Hard linking")?;
+            }
+        }
+    }
+    assert!(pathbuf.pop());
+    Ok(())
+}
+
+/// The common container storage model is to store blobs (layers) as unpacked directories,
+/// and use the Linux `overlayfs` to merge them dynamically.
+///
+/// However, today the `ostree-prepare-root` model as used by ostree expects a final flattened
+/// filesystem tree.  (That said, there is work in ostree to use composefs, which will add
+/// a huge amount of flexibility)
+///
+/// This function is an "eager" implementation of computing the filesystem tree, implementing
+/// the same algorithm as overlayfs, including processing whiteouts.
```

Comment:
I know that `podman image mount` exists because we're using it in this code.

The answer to your question is in the paragraph right above this one, but I will expand it further.

Reply:
OK there's a bit more text now in the `podman_ostree.rs` module doc.